gnucash/import-export/aqb/gnc-flicker-gui.c

```diff
@@ -114,38 +114,37 @@ get_num (const char ch)
 }
 
 /* convert the bank challenge into the 5 bits for the flicker data */
-static char
-*flicker_data (const char *challenge)
+static guint
+flicker_data (const char *challenge)
 {
     /* bitfield is a clock bit and a 4-bit code with the bits reversed
        (bit 1 is the least significant and bit 4 the most
        so 0x1 is 1000 and 0x8 is 0001) */
-     static const _Bool bits[16][5] =
-     {
-           {0, 0, 0, 0, 0}, {0, 1, 0, 0, 0}, {0, 0, 1, 0, 0}, {0, 1, 1, 0, 0},
-           {0, 0, 0, 1, 0}, {0, 1, 0, 1, 0}, {0, 0, 1, 1, 0}, {0, 1, 1, 1, 0},
-           {0, 0, 0, 0, 1}, {0, 1, 0, 0, 1}, {0, 0, 1, 0, 1}, {0, 1, 1, 0, 1},
-           {0, 0, 0, 1, 1}, {0, 1, 0, 1, 1}, {0, 0, 1, 1, 1}, {0, 1, 1, 1, 1}
-     };
-
-     /* prepend synchronization identifier */
-     const char pre[] = {'0', 'F', 'F', 'F'};
-     size_t len = sizeof (pre) + strlen (challenge) + 1;
-     char* code = (char*)g_malloc0 (len);
-     memcpy (code, pre, sizeof (pre));
-     memcpy (code + sizeof (pre), challenge, strlen (challenge));
+    static const _Bool bits[16][5] =
+    {
+        {0, 0, 0, 0, 0}, {0, 1, 0, 0, 0}, {0, 0, 1, 0, 0}, {0, 1, 1, 0, 0},
+        {0, 0, 0, 1, 0}, {0, 1, 0, 1, 0}, {0, 0, 1, 1, 0}, {0, 1, 1, 1, 0},
+        {0, 0, 0, 0, 1}, {0, 1, 0, 0, 1}, {0, 0, 1, 0, 1}, {0, 1, 1, 0, 1},
+        {0, 0, 0, 1, 1}, {0, 1, 0, 1, 1}, {0, 0, 1, 1, 1}, {0, 1, 1, 1, 1}
+    };
+
+    /* prepend synchronization identifier */
+    char* code = g_strdup_printf ("0FFF%s", challenge);
+    guint challenge_length = strlen (code);
 
     /* Swap the position of the bits in pairs throughout the bank challenge
        (low-order nibble first). */
-     for (guint i = 0; i < len - 1; i += 2)
-     {
-         guint val1 = get_num (code[i]);
-         guint val2 = get_num (code[i+1]);
-
-         memcpy (&bitarray[i], bits[val2], sizeof(bits[val2]));
-         memcpy (&bitarray[i+1], bits[val1], sizeof(bits[val1]));
-     }
-     return code;
+    for (guint i = 0; i < challenge_length - 1; i += 2)
```

Comment:
I think this will leave off the last character as `strlen` doesn't count the terminal `'\0'`.

Reply:
@Bob-IT, @jralls: I have made bank transfers and the TAN is accepted.

I also changed the counter from "challenge_length-1" to "challenge_length-0" and the generation of the TAN still works.

Reply:
@CWehli So just need to change it to challenge_length, if that is correct I will do that tomorrow before I push.

Reply:
> I also changed the counter from "challenge_length-1" to "challenge_length-0" and the generation of the TAN still works.

It's interesting that it worked without the last character. It makes one wonder if there's some padding included in the challenge.